src/impl_methods.rs

```diff
@@ -2473,6 +2473,45 @@ where
         self
     }
 
+    /// Consume the array, call `f` by **v**alue on each element, and return an
+    /// owned array with the new values. Works for **any** `F: FnMut(A)->B`.
+    ///
+    /// If `A` and `B` are the same type then the map is performed by delegating
+    /// to [`mapv_into()`] and then converting into an owned array. This avoids
+    /// unnecessary memory allocations in [`mapv()`].
+    ///
+    /// If `A` and `B` are different types then a new array is allocated and the
+    /// map is performed as in [`mapv()`].
+    ///
+    /// Elements are visited in arbitrary order.
+    pub fn mapv_into_any<B, F>(self, mut f: F) -> Array<B, D>
+    where
+        S: DataMut,
+        F: FnMut(A) -> B,
+        A: Clone + 'static,
+        B: 'static,
+    {
+        if core::any::TypeId::of::<A>() == core::any::TypeId::of::<B>() {
+            // A and B are the same type.
+            // Wrap f in a closure of type FnMut(A) -> A .
+            let f = |a| {
+                let b = f(a);
+                // Safe because A and B are the same type.
+                unsafe { core::ptr::read(&*ManuallyDrop::new(b) as *const B as *const A) }
+            };
+            // Delegate to mapv_into() using the wrapped closure.
+            // Convert output to a uniquely owned array of type Array<A, D>.
+            let output = self.mapv_into(f).into_owned();
+            // Change the return type from Array<A, B> to Array<B, D>.
+            // Again, safe because A and B are the same type.
+            unsafe { core::ptr::read(&*ManuallyDrop::new(output) as *const Array<A, D> as *const Array<B, D>) }
```

Comment:
We happen to have a function for this in this file - called `unlimited_transmute`. :)

Reply:
Yeah, I noticed that method in the codebase used in the context of transmuting one dimensionality into another, but I wasn't sure whether or not I should rely on it.  Do you want me to refactor to use `unlimited_transmute()`?

The problem with `unlimited_transmute()` in this setting is the additional runtime check:

```rust
assert_eq!(size_of::<A>(), size_of::<B>());
```

Asserting that the size of the underlying types is equal is unnecessary because `mapv_into_any()` has already checked that the types themselves are equal (in which case their size is also equal).  The extra runtime assertion is expensive in this case because it will be called inside a tight loop on each element of the array being mapped.

Reply:
I agree about ensuring no runtime overhead. `size_of` is known by the compiler at compile time, and the assertion disappears, is compiled out, if it doesn't trigger in this case.

Reply:
You are correct, the [runtime assertion is optimized away](https://rust.godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(fontScale:14,fontUsePx:'0',j:1,lang:rust,selection:(endColumn:12,endLineNumber:5,positionColumn:12,positionLineNumber:5,selectionStartColumn:12,selectionStartLineNumber:5,startColumn:12,startLineNumber:5),source:'use+std::mem::%7BManuallyDrop,+size_of%7D%3B%0Ause+std::ptr%3B%0A%0A//+Type+your+code+here,+or+load+an+example.%0Apub+unsafe+fn+unlimited_transmute%3CA,+B%3E(a:+A)+-%3E+B+%7B%0A++++assert_eq!!(size_of::%3CA%3E(),+size_of::%3CB%3E())%3B%0A++++ptr::read(%26*ManuallyDrop::new(a)+as+*const+A+as+*const+B)%0A%7D%0A%0A//+If+you+use+%60main()%60,+declare+it+as+%60pub%60+to+see+it+in+the+output:%0Apub+fn+main()+%7B%0A++++let+a:+i32+%3D+42%3B%0A++++let+_b:+i32+%3D+unsafe+%7B+unlimited_transmute(a)+%7D%3B%0A%7D%0A'),l:'5',n:'0',o:'Rust+source+%231',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:r1530,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),fontScale:14,fontUsePx:'0',j:1,lang:rust,libs:!(),options:'-C++++++++++++++++opt-level%3D1',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'rustc+1.53.0+(Editor+%231,+Compiler+%231)+Rust',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compiler:1,editor:1,fontScale:14,fontUsePx:'0',wrap:'1'),l:'5',n:'0',o:'Output+of+rustc+1.53.0+(Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4) at `-C opt-level=1` or higher.  It is obviously not optimized out for debug builds with optimization turned off.

I'll refactor to use `unlimited_transmute()`.